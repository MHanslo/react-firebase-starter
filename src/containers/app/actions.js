
/*
 * App Actions
 *
 */

 import {
  LOGIN,
  LOGOUT
 } from './constants';

 /**
  * Dispatched when user authentication succeeds
  *
  * @param  {object} response Firebase auth response
  *
  * @return {object} Action with a type of LOGIN_REQUEST_SUCCESS with response
  */
 export function login(response) {
   return {
     type: LOGIN,
     response
   };
 }

 /**
 * Dispatches logout action
 *
 * @return {object} Action with a type of LOGOUT
 */
 export function logout() {
   return {
     type: LOGOUT
   };
 }
