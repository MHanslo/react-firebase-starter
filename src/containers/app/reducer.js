/*
 * App Reducer
 *
*/

import {
  LOGIN,
  LOGOUT
} from './constants';

const initialState = {
  user: null
};

export default function appReducer(state = initialState, action) {
  switch (action.type) {
    case LOGIN:
      return {
        ...state,
        user: action.response
      };
    case LOGOUT:
      return {
        ...state,
        user: null
      };
    default:
      return state;
  }
}
