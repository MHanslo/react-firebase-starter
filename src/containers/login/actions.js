
/*
 * Login Actions
 *
 */

 import { browserHistory } from 'react-router';
 import { firebaseAuth } from '../../firebase';

 import {
  LOGIN_REQUEST,
  LOGIN_REQUEST_SUCCESS,
  LOGIN_REQUEST_FAILURE
 } from './constants';

 import {
  login,
  logout
} from '../app/actions';

 /**
  * Dispatched when a user login request is initiated
  *
  * @return {object} Action with a type of LOGIN_REQUEST
  */
 export function loginRequest() {
   return {
     type: LOGIN_REQUEST
   };
 }

 /**
  * Dispatched when user authentication succeeds
  *
  * @param  {object} Action with a type of LOGIN_REQUEST_SUCCESS
  *
  */
 export function loginRequestSuccess() {
   return {
     type: LOGIN_REQUEST_SUCCESS
   };
 }

 /**
  * Dispatched when user authentication fails
  *
  * @param  {object} error The error
  *
  * @return {object} Action with a type of LOGIN_REQUEST_ERROR with error
  */
 export function loginRequestError(error) {
   return {
     type: LOGIN_REQUEST_FAILURE,
     error
   };
 }

 /**
 * Ends Firebase session, redirects to /login, and dispatches logout() action
 *
 */
 export function logoutAndRedirect() {
   return (dispatch) => {
     return firebaseAuth.signOut().then(() => {
       dispatch(logout());
       browserHistory.push('/');
     }, (error) => {
       // An error happened.
     });
   };
 }

 /**
  * Login button clicked
  *
  * @param  {string} email The user email
  * @param  {string} password The user password
  *
  */
 export function loginButtonClicked(email, password) {
   return (dispatch) => {
     dispatch(loginRequest());
     const promise = firebaseAuth.signInWithEmailAndPassword(email, password);
     return promise.then((response) => {
       dispatch(login({
         id: response.uid,
         displayName: response.displayName,
         photoURL: response.photoURL,
         emailVerified: response.emailVerified,
         email: response.email
       }));
       setTimeout(() => {
         browserHistory.push('/');
         dispatch(loginRequestSuccess());
       }, 1000);
     }).catch((error) => {
       dispatch(loginRequestError(error));
     });
   };
 }
