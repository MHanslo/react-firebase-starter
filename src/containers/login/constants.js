/*
 * Login Constants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'YourContainer/YOUR_ACTION_CONSTANT';
 */

export const LOGIN_REQUEST = 'LOGIN_REQUEST';
export const LOGIN_REQUEST_SUCCESS = 'LOGIN_REQUEST_SUCCESS';
export const LOGIN_REQUEST_FAILURE = 'LOGIN_REQUEST_FAILURE';
