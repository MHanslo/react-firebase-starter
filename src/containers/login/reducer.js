/*
 * Login Reducer
 *
*/

import {
  LOGIN_REQUEST,
  LOGIN_REQUEST_SUCCESS,
  LOGIN_REQUEST_FAILURE
} from './constants';

const initialState = {
  isAuthenticating: false,
  emailError: null,
  passwordError: null
};

export default function loginReducer(state = initialState, action) {
  switch (action.type) {
    case LOGIN_REQUEST:
      return {
        ...state,
        isAuthenticating: true,
        emailError: null,
        passwordError: null
      };
    case LOGIN_REQUEST_SUCCESS:
      return {
        ...state,
        isAuthenticating: false
      };
    case LOGIN_REQUEST_FAILURE:
      if (action.error.code === 'auth/wrong-password') {
        return {
          ...state,
          isAuthenticating: false,
          emailError: null,
          passwordError: action.error.message
        };
      }
      return {
        ...state,
        isAuthenticating: false,
        emailError: action.error.message,
        passwordError: null
      };
    default:
      return state;
  }
}
