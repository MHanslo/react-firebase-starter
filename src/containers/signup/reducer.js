
/*
 * Signup Reducer
 *
 */

import {
  SIGNUP_REQUEST,
  SIGNUP_REQUEST_SUCCESS,
  SIGNUP_REQUEST_FAILURE
} from './constants';

const initialState = {
  isAuthenticating: false,
  emailError: null,
  passwordError: null
};

export default function signupReducer(state = initialState, action) {
  switch (action.type) {
    case SIGNUP_REQUEST:
      return {
        ...state,
        isAuthenticating: true
      };
    case SIGNUP_REQUEST_SUCCESS:
      return {
        ...state,
        isAuthenticating: false
      };
    case SIGNUP_REQUEST_FAILURE:
      if (action.error.code === 'auth/weak-password') {
        return {
          ...state,
          isAuthenticating: false,
          emailError: null,
          passwordError: action.error.message
        };
      }
      return {
        ...state,
        isAuthenticating: false,
        emailError: action.error.message,
        passwordError: null
      };
    default:
      return state;
  }
}
