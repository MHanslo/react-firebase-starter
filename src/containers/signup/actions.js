
/*
 * Signup Actions
 *
 */

import { browserHistory } from 'react-router';
import { firebaseAuth } from '../../firebase';

import {
 SIGNUP_REQUEST,
 SIGNUP_REQUEST_SUCCESS,
 SIGNUP_REQUEST_FAILURE
} from './constants';

import { login } from '../app/actions';

/**
 * Dispatched when a user attempts to sign up
 *
 * @return {object} Action with a type of SIGNUP_REQUEST
 */
export function signupRequest() {
  return {
    type: SIGNUP_REQUEST
  };
}

/**
 * Dispatched when user sign up succeeds
 *
 * @return {object} Action with a type of SIGNUP_REQUEST_SUCCESS
 */
export function signupRequestSuccess() {
  return {
    type: SIGNUP_REQUEST_SUCCESS
  };
}

/**
 * Dispatched when user sign up fails
 *
 * @return {object} Action with a type of SIGNUP_REQUEST_ERROR with error
 */
export function signupRequestError(error) {
  return {
    type: SIGNUP_REQUEST_FAILURE,
    error
  };
}

/**
 * Sign up button clicked
 *
 * @param  {string} email The user email
 * @param  {string} password The user password
 *
 */
export function signupButtonClicked(email, password) {
  return function(dispatch) {
    dispatch(signupRequest());
    const promise = firebaseAuth.createUserWithEmailAndPassword(email, password);
    return promise.then((response) => {
      dispatch(login(response));
      setTimeout(() => {
        browserHistory.push('/');
        dispatch(signupRequestSuccess());
      }, 1000);
    }).catch((error) => {
      dispatch(signupRequestError(error));
    });
  };
}
