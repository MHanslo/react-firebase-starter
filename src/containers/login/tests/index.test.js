import React from 'react';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import * as types from '../constants';
import * as actions from '../actions';
import { LOGIN, LOGOUT } from '../../app/constants';
import reducer from '../reducer';
import Login from '../index';

jest.mock('react-router');
jest.mock('../../../firebase');
const mockStore = configureStore([ thunk ]);
const initalState = {
  isAuthenticating: false,
  emailError: null,
  passwordError: null
};
const store = mockStore({
  login: initalState
});

/**
 * Container render
 */
it('should render without errors', () => {
  const component = renderer.create(
    <Provider store={store}>
      <Login />
    </Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

/**
 * Container actions
 */
describe('Login actions', () => {
  it('should create an action when a user login request is initiated', () => {
    const expectedAction = {
      type: types.LOGIN_REQUEST
    };
    expect(actions.loginRequest()).toEqual(expectedAction);
  });

  it('should create an action when when user authentication succeeds', () => {
    const expectedAction = {
      type: types.LOGIN_REQUEST_SUCCESS
    };
    expect(actions.loginRequestSuccess()).toEqual(expectedAction);
  });

  it('should create an action when user authentication fails', () => {
    const error = {
      code: 'auth/wrong-password',
      message: 'Password is incorrect.'
    };
    const expectedAction = {
      type: types.LOGIN_REQUEST_FAILURE,
      error
    };
    expect(actions.loginRequestError(error)).toEqual(expectedAction);
  });

  it('should create an action that logs in a user', () => {
    const expectedAction = [{
      type: types.LOGIN_REQUEST
    }, {
      type: LOGIN,
      response: {
        displayName: undefined,
        email: undefined,
        emailVerified: undefined,
        id: undefined,
        photoURL: undefined
      }
    }];
    const email = 'email@email.com';
    const password = 'password';
    return store.dispatch(actions.loginButtonClicked(email, password))
     .then(() => {
       expect(store.getActions()).toEqual(expectedAction);
     });
  });

  it('should create an action that logs out a user', () => {
    const expectedAction = { type: LOGOUT };
    return store.dispatch(actions.logoutAndRedirect())
     .then(() => {
       expect(store.getActions()[2]).toEqual(expectedAction);
     });
  });
});

describe('Login reducer', () => {
  it('should return the initial state', () => {
    expect(
      reducer(undefined, {})
    ).toEqual(initalState);
  });

  it('should handle LOGIN_REQUEST', () => {
    expect(
      reducer({}, {
        type: types.LOGIN_REQUEST
      })
    ).toEqual({
      isAuthenticating: true,
      emailError: null,
      passwordError: null
    });
  });

  it('should handle LOGIN_REQUEST_SUCCESS', () => {
    expect(
      reducer({}, {
        type: types.LOGIN_REQUEST_SUCCESS
      })
    ).toEqual({
      isAuthenticating: false
    });
  });

  it('should handle LOGIN_REQUEST_FAILURE', () => {
    expect(
      reducer({}, {
        type: types.LOGIN_REQUEST_FAILURE,
        error: {
          code: 'auth/wrong-password',
          message: 'Wrong password provided.'
        }
      })
    ).toEqual({
      isAuthenticating: false,
      emailError: null,
      passwordError: 'Wrong password provided.'
    });
  });
});
