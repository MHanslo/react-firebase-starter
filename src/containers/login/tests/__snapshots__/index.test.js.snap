exports[`test should render without errors 1`] = `
<section
  className="hero is-medium is-primary is-bold">
  <div
    className="hero-body">
    <div
      className="container">
      <h1
        className="title">
        Sign In
      </h1>
      <div
        className="box">
        <form
          onSubmit={[Function]}>
          <div>
            <label
              className="label">
              Email
            </label>
            <div
              style={
                Object {
                  "marginBottom": "10px",
                }
              }>
              <input
                className="input"
                onChange={[Function]}
                placeholder="Email"
                type="text"
                value={undefined} />
              
            </div>
          </div>
          <div>
            <label
              className="label">
              Password
            </label>
            <div
              style={
                Object {
                  "marginBottom": "10px",
                }
              }>
              <input
                className="input"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value={undefined} />
              
            </div>
          </div>
          <div
            className="control">
            <div>
              <button
                className="button is-primary"
                onClick={undefined}>
                Sign in
              </button>
            </div>
          </div>
        </form>
      </div>
      <p
        className="has-text-centered">
        Dont have an account? 
      </p>
    </div>
  </div>
</section>
`;
